lib/Sema/CSSimplify.cpp

```diff
@@ -2717,42 +2717,47 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
   }
 
   /// The behavior limit to apply to a concurrency check.
-  auto getConcurrencyDiagnosticBehavior = [&](bool forSendable) {
-    // We can only handle the downgrade for conversions.
-    switch (kind) {
-    case ConstraintKind::Conversion:
-    case ConstraintKind::ArgumentConversion:
-      break;
+  auto getConcurrencyFixBehavior = [&](
+        bool forSendable
+      ) -> Optional<FixBehavior> {
+        // We can only handle the downgrade for conversions.
+        switch (kind) {
+        case ConstraintKind::Conversion:
+        case ConstraintKind::ArgumentConversion:
+          break;
 
-    default:
-      return DiagnosticBehavior::Unspecified;
-    }
+        default:
+          if (!shouldAttemptFixes())
+            return None;
+
+          return FixBehavior::Error;
+        }
 
-    // For a @preconcurrency callee outside of a strict concurrency context,
-    // ignore.
-    if (hasPreconcurrencyCallee(this, locator) &&
-        !contextRequiresStrictConcurrencyChecking(DC, GetClosureType{*this}))
-      return DiagnosticBehavior::Ignore;
+        // For a @preconcurrency callee outside of a strict concurrency
+        // context, ignore.
+        if (hasPreconcurrencyCallee(this, locator) &&
+            !contextRequiresStrictConcurrencyChecking(DC, GetClosureType{*this}))
+          return FixBehavior::Suppress;
 
-    // Otherwise, warn until Swift 6.
-    if (!getASTContext().LangOpts.isSwiftVersionAtLeast(6))
-      return DiagnosticBehavior::Warning;
+        // Otherwise, warn until Swift 6.
+        if (!getASTContext().LangOpts.isSwiftVersionAtLeast(6))
+          return FixBehavior::DowngradeToWarning;
 
-    return DiagnosticBehavior::Unspecified;
+        return FixBehavior::Error;
   };
 
   // A @Sendable function can be a subtype of a non-@Sendable function.
   if (func1->isSendable() != func2->isSendable()) {
     // Cannot add '@Sendable'.
     if (func2->isSendable() || kind < ConstraintKind::Subtype) {
-      if (!shouldAttemptFixes())
-        return getTypeMatchFailure(locator);
-
-      auto *fix = AddSendableAttribute::create(
-          *this, func1, func2, getConstraintLocator(locator),
-          getConcurrencyDiagnosticBehavior(true));
-      if (recordFix(fix))
+      if (auto fixBehavior = getConcurrencyFixBehavior(true)) {
```

Comment:
We have a pattern for this kind of a situation - a fix declares a static `attempt` method which returns a fix pointer. This way `getConcurrencyFixBehavior` could be turned into a implementation detail of `AddSendableAttribute` and `MarkGlobalActorFunction` and moved to CSFix.cpp and the code here becomes:

```
if (auto fix = AddSendableAttribute::attempt(*this, func1, func2, getConstraintLocator(locator))) {
  if (record(fix))
    return getTypeMatchFailure(locator);
} else {
    return getTypeMatchFailure(locator);
}
```